crates/search/src/project_search.rs

```diff
@@ -1259,54 +1293,166 @@ impl ProjectSearchBar {
+                                .map(str::to_string)
+                        }) {
+                            new_query
+                        } else {
+                            model.search_history_cursor.reset();
+                            String::new()
+                        }
+                    });
+                    search_view.set_query(&new_query, cx);
+                } else if search_view
+                    .included_files_editor
+                    .focus_handle(cx)
+                    .is_focused(cx)
+                {
+                    let new_included = search_view.model.update(cx, |model, cx| {
+                        if let Some(new_included) = model.project.update(cx, |project, _| {
+                            project
+                                .search_included_history_mut()
+                                .next(&mut model.search_included_history_cursor)
+                                .map(str::to_string)
+                        }) {
+                            new_included
+                        } else {
+                            model.search_included_history_cursor.reset();
+                            String::new()
+                        }
+                    });
+                    search_view.set_search_included(&new_included, cx);
+                } else if search_view
+                    .excluded_files_editor
+                    .focus_handle(cx)
+                    .is_focused(cx)
+                {
+                    let new_excluded = search_view.model.update(cx, |model, cx| {
+                        if let Some(new_excluded) = model.project.update(cx, |project, _| {
+                            project
+                                .search_excluded_history_mut()
+                                .next(&mut model.search_excluded_history_cursor)
+                                .map(str::to_string)
+                        }) {
+                            new_excluded
+                        } else {
+                            model.search_excluded_history_cursor.reset();
+                            String::new()
+                        }
+                    });
+                    search_view.set_search_excluded(&new_excluded, cx);
+                }
             });
         }
     }
 
-    fn previous_history_query(&mut self, _: &PreviousHistoryQuery, cx: &mut ViewContext<Self>) {
+    fn previous_history(&mut self, _: &PreviousHistory, cx: &mut ViewContext<Self>) {
         if let Some(search_view) = self.active_project_search.as_ref() {
             search_view.update(cx, |search_view, cx| {
-                if search_view.query_editor.read(cx).text(cx).is_empty() {
-                    if let Some(new_query) = search_view
-                        .model
-                        .read(cx)
-                        .project
+                if search_view.query_editor.focus_handle(cx).is_focused(cx) {
```

Comment:
You could significantly reduce the duplication here if instead of having 3 different functions (search_history, search_included_history, search_excluded_history) we've had just one `search_history` that took Included/Excluded/Query as an argument.

Reply:
thanks